### PR TITLE
Added the basic machinery to submit metrics for nurses to DHIS2

### PR DESCRIPTION
**Pull request by chrislombaard**

### CODE REVIEW

File: nurseconnect/tasks.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from __future__ import absolute_import
+
+import requests
```
Comment (phala) [inline nurseconnect/tasks.py:3]: Fix order of imports

File: nurseconnect/tasks.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+from celery.schedules import crontab
+from celery.task import periodic_task
+from django.contrib.auth.models import User
+
+from raven.utils import json
+
+
+def nurses_registered():
+    """Returns the number of nurses registered on the system."""
+    return User.objects.all().count()
```
Comment (phala) [inline nurseconnect/tasks.py]: Wouldn't this take into account the superusers and staff users as well? You probably need some sort of filtering here

File: nurseconnect/tasks.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@periodic_task(
+    run_every=crontab(),
```
Comment (phala) [inline nurseconnect/tasks.py]: Are we sure we want this task to run every minute?

File: nurseconnect/tasks.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+from __future__ import absolute_import
+
+import requests
+from celery.schedules import crontab
+from celery.task import periodic_task
+from django.contrib.auth.models import User
+
+from raven.utils import json
```
Comment (phala) [inline nurseconnect/tasks.py]: Why not the standard lib's `json`?

File: nurseconnect/__init__.py [hunk 5]
```diff
@@ -1,3 +1,4 @@
+from __future__ import absolute_import
```
Comment (phala) [inline nurseconnect/__init__.py]: Why do we need this here?

File: nurseconnect/urls.py [hunk 6]
```diff
@@ -151,3 +150,9 @@
     urlpatterns += static(
         settings.MEDIA_URL + "images/",
         document_root=os.path.join(settings.MEDIA_ROOT, "images"))
```
Comment (phala) [inline nurseconnect/urls.py]: Can we not move that last closing bracket to the next line, to stick to a single convention?

File: nurseconnect/urls.py [hunk 7]
```diff
@@ -151,3 +150,9 @@
...
+    urlpatterns += \
+        url(
+            r"^styleguide/",
+            include("styleguide.urls",
+                    namespace="styleguide")
```
Comment (phala) [inline nurseconnect/urls.py]: Does PEP8 complain with this being on one line?

**Review by phala — COMMENTED** [review]
Just that bit of feedback please

File: nurseconnect/tasks.py (hunk 1 above)
Comment (chrislombaard) [reply nurseconnect/tasks.py:3]: Future imports need to occur at the beginning of the file.

File: nurseconnect/tasks.py (hunk 4 above)
Comment (chrislombaard) [reply nurseconnect/tasks.py]: Fixed

File: nurseconnect/tasks.py (hunk 2 above)
Comment (chrislombaard) [reply nurseconnect/tasks.py]: Rightly so. Will add a filter to correct.

File: nurseconnect/tasks.py (hunk 3 above)
Comment (chrislombaard) [reply nurseconnect/tasks.py]: No, changed it to every 12 hours

File: nurseconnect/__init__.py (hunk 5 above)
Comment (chrislombaard) [reply nurseconnect/__init__.py]: Removed.

File: nurseconnect/urls.py (hunk 6 above)
Comment (chrislombaard) [reply nurseconnect/urls.py]: Yes, Moved.

**Comment by bruskiza** [comment]
@phalamag @chrislombaard what's the status of this please?

We need to get this going into Jembi's system ASAP please.

cc @fritzbrand

**Comment by smn** [comment]
@chrislombaard & @fritzbrand can we get an update on this please? None of us have the necessary context to judge whether or not this is complete or what's required to complete it. With what we know we're not in a position to finish this or see it merged. Please assist.

File: nurseconnect/tasks.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import requests
+
+from celery.schedules import crontab
+from celery.task import periodic_task
+
+from django.contrib.auth.models import User
+
+
+URL = "https://praekelt:praekelt@npr-staging.jembi.org:5000/ws/rest/v1/" \
+      "nurseconnectdataset"
```
Comment (smn) [inline nurseconnect/tasks.py]: sorry, this is really sloppy, credentials should not be in pull-requests. We'll have to notify Jembi.

**Comment by phala** [comment]
@smn can you please review this?

In addition the the `CLINIC_CODE_API` environment variable, we'll also need
`JEMBI_URL`, `JEMBI_USERNAME` and `JEMBI_PASSWORD` to be set.

File: nurseconnect/tasks.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+def nurses_registered():
+    """Returns the number of nurses registered on the system."""
+    num_nurses = User.objects.filter(is_staff=False).count()
+    data = {
+        "dataValues": [
+            {
+                "dataElement": "CSv1k6HyWaX",
+                "period": "201601",
+                "orgUnit": "Fws0A9spb9F",
+                "value": str(num_nurses)
```
Comment (smn) [inline nurseconnect/tasks.py:41]: @phalamag are these needing to be fixed values?

**Comment by smn** [comment]
@phalamag thanks for helping with this, can we get a test for it?

File: nurseconnect/tasks.py (hunk 9 above)
Comment (phala) [reply nurseconnect/tasks.py:41]: @smn perhaps it could be better to have them saved as `constants` elsewhere.
I'll modify

**Comment by phala** [comment]
@smn sure. I'll write tests :)

File: nurseconnect/tasks.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def nurses_registered():
+    """Returns the number of nurses registered on the system."""
+    num_nurses = User.objects.filter(is_staff=False).count()
+    data = {
+        "dataValues": [
+            {
+                "dataElement": JEMBI["num_nurses"]["dataElement"],
+                "period": "201601",
```
Comment (erikh360) [inline nurseconnect/tasks.py]: This looks like a date that is hard coded, or is it something else?

File: nurseconnect/tasks.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    # nurses_per_facility = {"clinic_code": total} pairs
+    for k, v in nurses_per_facility.items():
+        if k:
+            clinic = get_clinic_code(k)
+            data = {
+                "dataValues": [
+                    {
+                        "dataElement": JEMBI["nurses_per_facility"][
+                            "dataElement"],
+                        "period": "201601",
```
Comment (erikh360) [inline nurseconnect/tasks.py]: Same as above

**Review by erikh360 — COMMENTED** [review]
It's just these two periods that I'm concerned about but other than that I'm happy.

File: nurseconnect/tasks.py (hunk 11 above)
Comment (phala) [reply nurseconnect/tasks.py]: I'm actually not sure what the real date should be here. I would assume its `yyyymm`. However, I think this is something @chrislombaard can probably clarify on when he picks up the rest of this work.

Comment (bruskiza) [reply nurseconnect/tasks.py]: @erikh360 @phalamag @chrislombaard:

* The tasks knows what date it's reporting for
* The date should be for what ever period it's reporting on

E.g.

Task runs on 1 March 2017 12am for Feb: period = 201703
Task runs on 1 Dec 2016 for Nov: period = 201611

Comment (chrislombaard) [reply nurseconnect/tasks.py]: Yes, @erikh360 @bruskiza @phalamag That date is a placeholder and will need to be set based on when the task is run. I'll look into this today.

File: nurseconnect/tests/test_tasks.py [hunk 12]
```diff
@@ -1,5 +1,7 @@
 import responses
 
+from datetime import datetime
```
Comment (phala) [inline nurseconnect/tests/test_tasks.py:1]: This should right at the top of the imports

File: nurseconnect/tasks.py [hunk 13]
```diff
@@ -62,7 +64,8 @@ def nurses_registered_per_clinic():
                     {
                         "dataElement": JEMBI["nurses_per_facility"][
                             "dataElement"],
-                        "period": "201601",
+                        "period": str(datetime.now().year) +
+                        "%02d" % datetime.now().month,
```
Comment (phala) [inline nurseconnect/tasks.py]: Can this not be indented?

File: nurseconnect/tasks.py [hunk 14]
```diff
@@ -35,7 +36,8 @@ def nurses_registered():
         "dataValues": [
             {
                 "dataElement": JEMBI["num_nurses"]["dataElement"],
-                "period": "201601",
+                "period": str(datetime.now().year) +
+                "%02d" % datetime.now().month,
```
Comment (phala) [inline nurseconnect/tasks.py]: Can this not be indented?

File: nurseconnect/tests/test_tasks.py (hunk 12 above)
Comment (chrislombaard) [reply nurseconnect/tests/test_tasks.py:1]: Gotcha

**Review by erikh360 — APPROVED** [review]
+1

**Comment by phala** [comment]
+1

**Pull request merged**